### PR TITLE
tsdb/agent: allow ingestion of OOO samples

### LINKED ISSUE #12673: Unable to send Out of Order Metrics from Agent (and Grafana Agent)
### What did you do?

This code https://github.com/prometheus/prometheus/blob/4a56a6bf592ad032714ddedaee1b5e4fc77aa93e/tsdb/agent/db.go#L805 looks to block any out of order samples from being sent. Even if the endpoint allows out of order. 

### What did you expect to see?

Samples out of order to be determined by the endpoint, or at least configurable.

### What did you see instead? Under which circumstances?

Samples blocked early.

### System information

_No response_

### Prometheus version

_No response_

### Prometheus configuration file

_No response_

### Alertmanager version

_No response_

### Alertmanager configuration file

_No response_

### Logs

_No response_

### PR DESCRIPTION
**Pull request by tpaschalis**
After a [discussion](https://cloud-native.slack.com/archives/C01AUBA4PFE/p1695122956103869) in the CNCF Slack's #prometheus-dev channel, I'm opening this PR to enable OOO ingestion of samples for the Prometheus Agent.

I've also added a test for this new behaviour. The test is more complicated than expected, given that we have to replay the WAL to read the lastTs for a series, let me know if you think we're better off without it.

Closes #12673.

### CODE REVIEW

File: tsdb/agent/db_test.go [hunk 1]
```diff
@@ -751,3 +751,119 @@ func TestStorage_DuplicateExemplarsIgnored(t *testing.T) {
...
+	for _, l := range lbls {
+		lset := labels.New(l...)
+
+		floatHistograms := tsdbutil.GenerateTestFloatHistograms(numHistograms)
+
+		for i := 0; i < numDatapoints; i++ {
+			_, err := app.AppendHistogram(0, lset, int64(i), nil, floatHistograms[i])
+			require.NoError(t, err)
+		}
+	}
```
Comment (jesusvazquez) [inline tsdb/agent/db_test.go:868]: @tpaschalis So far I see that we are checking that there are no erros on append but we are not checking that the samples are actually there. Could we test to replay and read that there is what we actually inserted?

Comment (tpaschalis) [reply tsdb/agent/db_test.go:873]: Hmm, the thing is that there's no way to access the samples by replaying the WAL (they're only used to recalculate lastTs).

What we _can_ do is append a different number of samples and use the `prometheus_agent_samples_appended_total` metric to verify that the right number of datapoints was appended. I've implemented this in 242aa5706d82c290294c34c27bd6bca539edb4a3, let me know how it looks!

Comment (jesusvazquez) [reply tsdb/agent/db_test.go:873]: That works fine :+1:

**Review by jesusvazquez — APPROVED** [review]

**Comment by datsabk** [comment]
@tpaschalis - How is this different from the experimental feature of out-of-order ingestion that was already enabled? 

Trying to understand if this is a new feature that's introduced or just a past bug that got solved?

**Pull request merged**